service/share/cache_availability.go

```diff
@@ -0,0 +1,64 @@
+package share
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/autobatch"
+	"github.com/ipfs/go-datastore/namespace"
+)
+
+var (
+	// DefaultWriteBatchSize defines the size of the batched header write.
+	// Headers are written in batches not to thrash the underlying Datastore with writes.
+	// TODO(@Wondertan, @renaynay): Those values must be configurable and proper defaults should be set for specific node
+	//  type. (#709)
+	DefaultWriteBatchSize = 2048
```

Comment:
So what happens if in between 2048 sample requests, the node goes down in between? Worst case scenario the root does not get stored and we resample, right? Not too bad I guess. But 2048 headers might be a bit too much. Also it is not really headers that are batched her but actually just root hashes, right?

Reply:
Yep, we are batching here hashes of dahs that were dased. The number is ok.

Reply:
I believe the underlying datastore that gets passed to `cacheAvailability`, upon stop, gets closed which flushes the queued up writes to disk. This is true, right @Wondertan ? Otherwise we should probably add the datastore's "lifecycle" to the node? Or i can make a hook for flushing writes on stop inside the constructor.

Reply:
Yep, it is getting closed. The problem is with auto batch which is not getting closed and it wraps our datastore on the node. So we should close the auto batch first, so it can write all the outstanding writes to the datastore and then close it as well. This can be simply done by the Stop on cacheAvailability which stops the autobatch. The ordering is handled by fx